dynamodb/schema.go

```diff
@@ -0,0 +1,358 @@
+	typeBool           = "Bool"
+	typeNumberInt      = "NumberInt"
+	typeNumberFloat    = "NumberFloat"
+	typeBinary         = "Binary"
+	typeList           = "List"
+	typeMap            = "Map"
+	typeStringSet      = "StringSet"
+	typeNumberIntSet   = "NumberIntSet"
+	typeNumberFloatSet = "NumberFloatSet"
+	typeBinarySet      = "BinarySet"
+
+	errThreshold      = float64(0.001)
+	conflictThreshold = float64(0.05)
+)
+
+type dynamoClient interface {
+	ListTables(input *dynamodb.ListTablesInput) (*dynamodb.ListTablesOutput, error)
+	DescribeTable(input *dynamodb.DescribeTableInput) (*dynamodb.DescribeTableOutput, error)
+	Scan(input *dynamodb.ScanInput) (*dynamodb.ScanOutput, error)
+}
+
+// ProcessSchema performs schema conversion for source tables in a DynamoDB
+// database. We use the standard APIs to obtain source table's schema
+// information, including primary keys and secondary indexes. DynamoDB is a
+// schemaless database that some optional attributes can be missed or has
+// different data types in some rows. Therefore, we have to sample a number of
+// rows to infer the schema for optional attributes.
+func ProcessSchema(conv *internal.Conv, client dynamoClient, tables []string, sampleSize int64) error {
+	if len(tables) == 0 {
+		var err error
+		tables, err = listTables(client)
+		if err != nil {
+			return err
+		}
+		if len(tables) == 0 {
+			return fmt.Errorf("no DynamoDB table exists under this account")
+		}
+	}
+	for _, t := range tables {
+		if err := processTable(conv, client, t, sampleSize); err != nil {
+			return err
+		}
+	}
+	schemaToDDL(conv)
+	conv.AddPrimaryKeys()
+	return nil
+}
+
+func listTables(client dynamoClient) ([]string, error) {
+	var tables []string
+	var lastEvaluatedTableName *string
+
+	for {
+		input := &dynamodb.ListTablesInput{}
+		if lastEvaluatedTableName != nil {
+			input.ExclusiveStartTableName = lastEvaluatedTableName
+		}
+
+		result, err := client.ListTables(input)
+		if err != nil {
+			return nil, err
+		}
+		for _, t := range result.TableNames {
+			tables = append(tables, *t)
+		}
+
+		if result.LastEvaluatedTableName == nil {
+			return tables, nil
+		}
+		lastEvaluatedTableName = result.LastEvaluatedTableName
+	}
+}
+
+func processTable(conv *internal.Conv, client dynamoClient, table string, sampleSize int64) error {
+	dySchema := dynamoDBSchema{TableName: table}
+	err := dySchema.parseIndexes(client)
+	if err != nil {
+		return err
+	}
+	stats, count, err := dySchema.scanSampleData(client, sampleSize)
+	if err != nil {
+		return err
+	}
+	dySchema.inferDataTypes(stats, count)
+	conv.SrcSchema[table] = dySchema.genericSchema()
+	return nil
+}
+
+type dynamoDBSchema struct {
+	TableName   string
+	ColumnNames []string
+	ColumnTypes map[string]colType
+	PrimaryKeys []string
+	SecIndexes  []index
+}
+
+type colType struct {
+	Name     string
+	Nullable bool
+}
+
+type index struct {
+	Name string
+	Keys []string
+}
+
+func (s *dynamoDBSchema) parseIndexes(client dynamoClient) error {
```

Comment:
Maybe analyzeMetadata instead of parseIndexes?
I'm not sure DynamoDB considers primary keys to be indexes?